### DISCUSSION TITLE
Json

### DISCUSSION THREAD

**Discussion by avico78** [collerek/ormar#182] [Q&A]
I'd like to understand what are the limitation for presenting a Json 
with ormar .

The use case : 
have requirement for specific  Json structure required ,
the input data lives in tables ,
The tables build in with structure and relationship to other tables that can be reflect in a Json.

Cases i couldn't find to solve : 

1.dictionary object in Json
i.e:
```
{
"address": {"address": "some address","zip code"}
}
```
seems field in ormar allow using a dict but it doesn't reflect in json:  
```python


class Address(AddressBase):
    class Meta(BaseMetaSyntSrc):
        tablename: str = "customer_address"
    country: dict = ormar.String(max_length=40, default = lambda : fake.country())
```
second case- the a Dictrionary object related to different table 
Meaning in Json it look the same , im just fetching the data from different table.

```python

class Address(AddressBase):
    class Meta(BaseMetaSyntSrc):
        tablename: str = "customer_address"
    country: Optiona[Dict[Customer]] = ormar.ForeignKey(Customer,related_name="subscriber")
````
Far as i remember , when using conventional pydatnic class this was possible.

2.Json also allow using "$" in filed name,
is there an option presenting such filed with "$" ?
i.e:
```
{
$: "Werido but legal"
}
```
3.null/Null values

How can set field to have "constant" null in Json,
meaning is mandatory  regardless if it has or has not any value.

i.e:
```
{
"filed: null
}
```
in addition , in many cases we want to minimize Json size by alimenting non mandatory field having null value from Json ,
meaning if it define in the class but if there's no data it - suppressing it from the Json .

**Comment by collerek** [comment]
1.
Ormar has Json field type that can hold dicts or lists or whatever a valid Json is. Ormar will take care of loading/dumping it to string.

As for taking data from another table I'm not sure I understand what you would like to do, can you explain in more details with sample use case?

2.
I think string "$" is a valid python key and you san include it in json field.

3.
In dicts you can do whatever you want, in model fields if you set fields to optional or set default to none it should still appear in dict and json

4.
Pydantic dict has flags `exclude_default` and `exclude_unset` (or something similar, I'm writing from a phone so it's hard to check now, check pydanric docs). First one exclude values if the value is not passed and a default is used, second one excludes fields with None.

Now i didn't test those two with ormar so I'm not sure how it handles ormar defaults etc. But you can give it a try, i will check that when implementing exclude_primary_keys issue #164
